repose-aggregator/components/filters/header-normalization/src/main/scala/org/openrepose/filters/headernormalization/HeaderNormalizationFilter.scala

```diff
@@ -0,0 +1,133 @@
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+
+package org.openrepose.filters.headernormalization
+
+import java.net.URL
+import java.util.concurrent.TimeUnit
+import javax.inject.{Inject, Named}
+import javax.servlet._
+import javax.servlet.http.HttpServletRequest
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.config.manager.UpdateListener
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.openrepose.core.filter.FilterConfigHelper
+import org.openrepose.core.filters.HeaderNormalization
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.services.reporting.metrics.MetricsService
+import org.openrepose.core.services.reporting.metrics.impl.MeterByCategorySum
+import org.openrepose.filters.headernormalization.config.{HttpMethod, HeaderNormalizationConfig}
+
+import scala.collection.JavaConverters._
+import scala.util.matching.Regex
+
+@Named
+class HeaderNormalizationFilter @Inject()(configurationService: ConfigurationService, metricsService: MetricsService)
+  extends Filter with UpdateListener[HeaderNormalizationConfig] with LazyLogging {
+
+  import HeaderNormalizationFilter._
+
+  private var configurationFile: String = DEFAULT_CONFIG
+  private var initialized = false
+  private var config: Seq[Target] = _
+  private var metricsMeter: Option[MeterByCategorySum] = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("Header Normalization filter initializing...")
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing Header Normalization filter using config $configurationFile")
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[HeaderNormalizationConfig])
+
+    metricsMeter = Option(metricsService) match {
+      case Some(service) =>
+        logger.debug("Initializing metrics meter for Header Normalization filter")
+        Some(service.newMeterByCategorySum(classOf[HeaderNormalization], "header-normalization", "Normalization", TimeUnit.SECONDS))
+      case None =>
+        logger.warn("Metrics meter not started for Header Normalization filter due to missing metrics service")
+        None
+    }
+
+    logger.trace("Header Normalization filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+
+    config.find { target =>
+      // find the first "target" config element that matches this request (if any)
+      target.url.findFirstIn(wrappedRequest.getRequestURL).isDefined &&
```

Comment:
It looks like the filter used to do:
```
target.getPattern().matcher(uri).matches()
```

The semantics of .findFirstIn for a Java regex are not the same as the semantics of Java's Pattern .matches(). The former will check if any substring of the input matches the regex and return the first such match, while the latter will require that the full input matches the input.

Example:
```
scala> "hello".r.findFirstIn("hello world!")
res0: Option[String] = Some(hello)

scala> "hello".r.pattern.matcher("hello world!").matches()
res1: Boolean = false
```

I'd recommend changing this line to:
```
target.url.pattern.matcher(wrappedRequest.getRequestURL).matches() &&
```